### DISCUSSION TITLE
Prisma with hashids

### DISCUSSION THREAD

**Discussion by lukecarr** [prisma/prisma#10588] [Q&A] [answered]
Does anyone have any experience implementing hashid'd autoincrementing IDs with Prisma?

We have a schema that uses autoincrementing IDs throughout but would like to use [hashids](https://hashids.org/) to encode/decode IDs when they become "public facing".

Right now, we have a rather hacky internal library that tracks all of our known ID property names and encodes/decodes them automatically, but it's becoming quite cumbersome to use.

However, if there isn't anything out there, I'd love to sit down and refactor our internal library into something that could be open-sourced.

**Comment by austincrim** [comment]
Hey @lukecarr 👋,

I don't have any experience with hashids specifically, but it sounds like you might want to look at our [middleware API](https://www.prisma.io/docs/concepts/components/prisma-client/middleware). It offers quite a bit of flexibility and I think you could come up with a solution whether you want to store the hashids or store plain IDs and hash them on read.

Let me know if I can answer any more specific questions you have!

**Reply by lukecarr** [reply]
Hi @austincrim 

How have I not come across this yet; middlewares solve so many of our gripes with Prisma!

Having a brief look over the API, it looks like I'll need to inspect `params.action` so hashid **encoding** happens on read ops, and **decoding** happens on write ops.

Related question: are you aware of any way to introspect the schema at runtime to determine the primary/foreign key (and which model the foreign key relates to) field names for each model?

**Reply by austincrim** [reply]
> Related question: are you aware of any way to introspect the schema at runtime to determine the primary/foreign key (and which model the foreign key relates to) field names for each model?

We have an undocumented `Primsa.dmmf` API on the `Prisma` import. 
```ts
import { Prisma } from '@prisma/client'

console.log(Prisma.dmmf)
```

It contains Prisma's internal representation of the parsed schema and should have all the information you need. Again, it is **undocumented**, so your experience using it might not be ideal.

What is your use case for needed this information? Maybe you could open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with your needs?

**Reply by lukecarr** [reply, marked as answer]
> What is your use case for needed this information?

We're looking to implement an ID system similar to [Stripe's](https://gist.github.com/fnky/76f533366f75cf75802c8052b577e2a5) where each model has a unique PK prefix.

However, we'd still like to use auto-incrementing IDs at the database level, so right now we're using [hashids](https://hashids.org/) at the application layer to encode/decode each model's ID fields (and prepend the correct prefix). Naturally, we also need to encode/decode any FK fields for a model in other models.

Right now, are internal lib comprises of a register to keep track of all models, their PK prefixes, and their FK fields in other models. It also has an encode/decode function for transforming an object (looking at the register and encoding/decoding 

> Maybe you could open a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) with your needs?

I think it's a bit too specific for our scenario to warrant a feature request; from what I can gather, middleware and `Prisma.dmmf` should resolve all of our issues. Happy to open a feature request though if you think otherwise! 😄

**Reply by lararosekelley** [reply]
Found this thread, we have a similar problem where I work - we use `hashids` as well, and ideally would be able to only work with those string IDs in our application code, while still preserving the autoincrementing integer IDs at the database layer. Curious if you came to middleware or `dmmf`-based solution to allow passing strings to Prisma's various functions, e.g.

```typescript
const foo = prisma.foo.findUniqueOrThrow({ where: { id: "some string" });
```

Otherwise, we're kinda stuck calling `encode` and `decode` functions all over the place to satisfy Prisma's need for these Ids to all be numbers

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
